src/sinks/util/auto_concurrency/controller.rs

```diff
@@ -0,0 +1,189 @@
+use futures::ready;
+use std::cmp::max;
+use std::future::Future;
+use std::mem::{drop, replace};
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+use std::time::{Duration, Instant};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+
+const EWMA_ALPHA: f64 = 0.5;
+const THRESHOLD_RATIO: f64 = 0.01;
+
+#[derive(Clone, Copy, Debug, Default)]
+struct EWMA {
+    average: f64,
+}
+
+impl EWMA {
+    fn average(&self) -> f64 {
+        self.average
+    }
+
+    fn update(&mut self, point: f64) -> f64 {
+        self.average = match self.average {
+            avg if avg == 0.0 => point,
+            avg => point * EWMA_ALPHA + avg * (1.0 - EWMA_ALPHA),
+        };
+        self.average
+    }
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+struct Mean {
+    sum: f64,
+    count: usize,
+}
+
+impl Mean {
+    fn update(&mut self, point: f64) -> f64 {
+        self.sum += point;
+        self.count += 1;
+        // Return current average
+        self.sum / max(self.count, 1) as f64
+    }
+
+    fn reset(&mut self) {
+        self.sum = 0.0;
+        self.count = 0;
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum ResponseType {
+    Normal,
+    BackPressure,
+    Other,
+}
+
+pub(crate) trait IsBackPressure {
+    fn is_back_pressure(&self) -> bool;
+}
+
+/// Shared class for `tokio::sync::Semaphore` that manages adjusting the
+/// semaphore size and other associated data.
+#[derive(Debug)]
+pub(super) struct Controller {
+    semaphore: Arc<Semaphore>,
+    max: usize,
+    inner: Arc<Mutex<Inner>>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    current: usize,
+    to_forget: usize,
+    past_rtt: EWMA,
+    next_update: Instant,
+    current_rtt: Mean,
+    had_back_pressure: bool,
+}
+
+impl Controller {
+    pub(super) fn new(max: usize, current: usize) -> Self {
+        Self {
+            semaphore: Arc::new(Semaphore::new(current)),
+            max,
+            inner: Arc::new(Mutex::new(Inner {
+                current: current,
+                to_forget: 0,
+                past_rtt: Default::default(),
+                next_update: Instant::now(),
+                current_rtt: Default::default(),
+                had_back_pressure: false,
+            })),
+        }
+    }
+
+    pub(super) fn acquire(&self) -> impl Future<Output = OwnedSemaphorePermit> + Send + 'static {
+        MaybeForgetFuture {
+            semaphore: self.semaphore.clone(),
+            inner: self.inner.clone(),
+            future: Box::pin(Arc::clone(&self.semaphore).acquire_owned()),
+        }
+    }
+
+    pub(super) fn adjust_to_response<T, E>(&self, start: Instant, response: &Result<T, E>)
+    where
+        E: IsBackPressure,
+    {
+        let response = match response {
+            Ok(_) => ResponseType::Normal,
+            Err(r) if r.is_back_pressure() => ResponseType::BackPressure,
```

Comment:
An alternative, as you suggest, may be to extend the retry logic type to include the needed functionality (if it doesn't already) and use that instead of a new type. It might still need additional trait object, but at least there wouldn't be the need to implement another trait.